chain/client/src/chunks_delay_tracker.rs

```diff
@@ -1,192 +1,103 @@
-        let heights_to_drop: Vec<BlockHeight> = self
-            .heights
-            .range(..ChunksDelayTracker::lowest_height(head_height))
-            .map(|(&k, _v)| k)
-            .collect();
-        for h in heights_to_drop {
-            self.heights.remove(&h);
-        }
+    pub fn received_chunk(&mut self, chunk_hash: &ChunkHash, timestamp: Instant) {
+        self.chunks_in_progress
+            .get_mut(&chunk_hash)
+            .map(|chunk_in_progress| chunk_in_progress.chunk_received.get_or_insert(timestamp));
     }
 
-    // Computes the delay between receiving a block and receiving the latest of its chunks.
-    // Updates the metric to the max delay across the most recently processed blocks.
-    fn get_max_delay(&mut self, head_height: BlockHeight) -> Duration {
-        let mut max_delay = Duration::ZERO;
-        for (_, v) in self.heights.range(..=head_height) {
-            if let Some(block_received) = v.block_received {
-                if let Some(latest_chunk) = v.chunks_received.values().max() {
-                    if latest_chunk > &block_received {
-                        let delay = latest_chunk.duration_since(block_received);
-                        max_delay = max(max_delay, delay);
-                    }
-                }
-            }
-        }
-        max_delay
+    pub fn requested_chunk(&mut self, chunk_hash: &ChunkHash, timestamp: Instant) {
+        self.chunks_in_progress.entry(chunk_hash.clone()).or_insert_with(|| ChunkInProgress {
+            chunk_requested: timestamp,
+            chunk_received: None,
+        });
     }
 
-    fn update_chunks_metric(&mut self, head_height: BlockHeight) {
-        metrics::CHUNKS_RECEIVING_DELAY_US.set(self.get_max_delay(head_height).as_micros() as i64);
-    }
+    pub fn finish_block_processing(
+        &mut self,
+        processed_block_hash: &CryptoHash,
+        chunks: &[ChunkHash],
+    ) {
+        self.update_block_chunks_requested_metric(processed_block_hash, &chunks);
+        self.update_chunks_metric(&chunks);
 
-    // Computes the difference between the latest block we are aware of and the current head.
-    fn get_blocks_ahead(&mut self, head_height: BlockHeight) -> u64 {
-        if let Some((&latest, _)) = self.heights.iter().next_back() {
-            latest.saturating_sub(head_height)
-        } else {
-            0
+        self.blocks_received.remove(&processed_block_hash);
+        for chunk_hash in chunks {
+            self.chunks_in_progress.remove(chunk_hash);
         }
     }
 
-    fn update_blocks_ahead_metric(&mut self, head_height: BlockHeight) {
-        metrics::BLOCKS_AHEAD_OF_HEAD.set(self.get_blocks_ahead(head_height) as i64);
-    }
-
-    fn update_metrics(&mut self, head_height: BlockHeight) {
-        self.update_chunks_metric(head_height);
-        self.update_blocks_ahead_metric(head_height);
+    pub fn get_chunks(block: &Block) -> Vec<ChunkHash> {
```

Comment:
nit: I find it strange to see this method here (as it has nothing to do with chunks_delay_tracking) - maybe we could inline it in that place where it is used?

Reply:
Done